### PR TITLE
Ansible 3.1.0 & ansible-lint 5.0.4

### PR DESCRIPTION
**Pull request by laserscout**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by carlocab** [comment]
Cherry-picked #73430 for inclusion here to resolve the test failure.

I'm kinda bothered by both `ansible-lint` and `ansible` needing an `ansible-base` resource. I wonder if there's a way to fix this so that only `ansible` needs it.

**Comment by laserscout** [comment]
Yup, I'm still working on it; will start by following the direction you pointed you here

> ...
> Not sure if it's necessary to have an `ansible-base` formula. It might be better to just install `ansible-base` into the global environment (instead of the formula-specific venv) in the `ansible` formula. Then, formulae that need `ansible-base` (e.g. `ansible-lint`) can just depend on `ansible`.
> 
> To install `ansible-base` into the global environment, we can do something like this:
> 
> ```ruby
> resource("ansible-base").stage do
>   system "python3", *Language::Python.setup_install_args(prefix)
> end
> ```
> 
> We'd have to exclude `ansible-base` from the resources that are installed into the `ansible`-specific venv, but that's not too hard to do. (See, for example, formulae that use the resource `Pillow`.)

_Originally posted by @carlocab in https://github.com/Homebrew/homebrew-core/issues/73265#issuecomment-803276047_

**Comment by carlocab** [comment]
Right. I forgot I wrote that. I guess one difficulty might be is that I _think_ doing that also requires installing `ansible-base`'s dependencies globally using
```ruby
system "python3", *Language::Python.setup_install_args(prefix)
```
too.

**Comment by laserscout** [comment]
Is it possible to not install it in a global path, but path that will be shared among `ansible-base` dependent formulae?

Say `(brew --prefix)/{lib,shared}/ansible`?

Then the path can be added to the python import path for the `ansible` and `ansible-lint` formulae

**Comment by carlocab** [comment]
Maybe, yes. Would it be enough to leave the `install` method of `ansible` unchanged, but fix `ansible-lint` so that its import path includes `ansible`'s `libexec`? Then we can remove all the resources from `ansible-lint` that are already in `ansible`.

**Comment by laserscout** [comment]
Doesn't the `libexec` path change after each update? That would mean an `ansible-lint` reinstallation would be required after each `ansible` update.

**Comment by carlocab** [comment]
It seems like they get version bumps at the same time anyway. That said, you can use `opt_libexec` for a stable path that doesn't change between version bumps.

**Comment by laserscout** [comment]
I removed all `ansible-base` related dependencies from the ansible-lint formula, and then added a `homebrew-ansible.pth` file pointing to `ansible`'s  `opt_libexec`

This does limit the `ansible-lint` formula to only work with the latest ansible version, installed by the `ansible` formula; while being perfectly capable in working with `ansible@2.9`.

I'm not very proficient in homebrew to even know if this sort of setup (allowing ansible-lint to depend, and work with either ansible versions) is even possible...

File: Formula/ansible-lint.rb [hunk 1]
```diff
@@ -124,6 +86,10 @@ class AnsibleLint < Formula
 
   def install
     virtualenv_install_with_resources
+    version = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
+    site_packages = "libexec/lib/python#{version}/site-packages"
```
Comment (carlocab) [inline Formula/ansible-lint.rb:90]:
```suggestion
    xy = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
    site_packages = "libexec/lib/python#{xy}/site-packages"
```
Let's not shadow the variable `version` defined in the formula.

Comment (laserscout) [reply Formula/ansible-lint.rb:90]: Yup

File: Formula/ansible-lint.rb [hunk 2]
```diff
@@ -124,6 +86,10 @@ class AnsibleLint < Formula
 
   def install
     virtualenv_install_with_resources
+    xy = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
+    site_packages = "libexec/lib/python#{xy}/site-packages"
+    ansible = Formula["ansible"].opt_prefix
```
Comment (carlocab) [inline Formula/ansible-lint.rb:91]:
```suggestion
    site_packages = "lib/python#{xy}/site-packages"
    ansible = Formula["ansible"].opt_libexec
```

File: Formula/ansible-lint.rb [hunk 3]
```diff
@@ -124,6 +86,10 @@ class AnsibleLint < Formula
...
+    (prefix/site_packages/"homebrew-ansible.pth").write ansible/site_packages
```
Comment (carlocab) [inline Formula/ansible-lint.rb:92]: Doesn't this need to be something like
```ruby
(prefix/site_packages/"homebrew-ansible.pth").write "import site; site.addsitedir('#{ansible/site_packages}')"
```
?
You can probably also simplify this a bit by defining
```ruby
ansible = Formula["ansible"].opt_libexec
```
instead.

Comment (laserscout) [reply Formula/ansible-lint.rb:92]: It doesn't seem necessary to execute `site.addsitedir()` in a `.pth` file [doc](https://docs.python.org/3/library/site.html).

Just adding the path worked for me, I could see the path when executing the venv python and ansible-lint did execute.

Comment (carlocab) [reply Formula/ansible-lint.rb:92]: Huh, weird. TIL.

File: Formula/ansible-lint.rb (hunk 2 above)
Comment (laserscout) [reply Formula/ansible-lint.rb:91]: I will use `Formula["ansible"].opt_libexec` and `libexec` instead of `prefix` for `ansible-lint`'s path

Comment (carlocab) [reply Formula/ansible-lint.rb:91]: Whoops, nope, this doesn't work.

Comment (laserscout) [reply Formula/ansible-lint.rb:91]: Worked!

**Comment by carlocab** [comment]
> This does limit the `ansible-lint` formula to only work with the latest ansible version, installed by the `ansible` formula; while being perfectly capable in working with `ansible@2.9`.

This is fine. If this doesn't work for someone, then they can put in the work to fix it. That said, I think even in its current form `ansible-lint` only seems to work with the latest `ansible` version, so I don't think this should cause a regression for anyone.

Thanks very much, @laserscout.

File: Formula/ansible.rb [hunk 4]
```diff
@@ -25,92 +25,38 @@ class Ansible < Formula
   uses_from_macos "libffi"
   uses_from_macos "libxslt"
 
-  # For instructions on how to update the resource blocks, see:
-  # https://github.com/Homebrew/homebrew-core/pull/50504#issuecomment-592743222
-
-  # Collect requirements from:
+  # The first, short block of resources can't be automatically updated and should be cared for manually.
+  # The second, long black can be updated by two slightly different methods.
```
Comment (carlocab) [inline Formula/ansible.rb:29]:
```suggestion
  # The second (short) block of resources can't be automatically updated and should be cared for manually.
  # The first (long) block can be updated by two slightly different methods.
```

**Comment by laserscout** [comment]
> This is fine. If this doesn't work for someone, then they can put in the work to fix it.

Yea, I'm one of these someone's which use ansible 2.9 =P.

> I think even in its current form `ansible-lint` only seems to work with the latest `ansible` version

It works fine with older versions

> Thanks very much, @laserscout.

A pleasure as always @carlocab

**Comment by carlocab** [comment]
> Yea, I'm one of these someone's which use ansible 2.9 =P.

Ha, well, we can work out how to get it to work with ansible@2.9 in another PR if you need it, then.

**Review by carlocab — APPROVED** [review]
Thanks again, @laserscout.

**Pull request closed without merging**